readme.md
[<img src="https://s3-us-west-2.amazonaws.com/arc.codes/architect-logo-500b@2x.png" width=500>](https://www.npmjs.com/package/@architect/functions)

## [`@architect/functions`](https://www.npmjs.com/package/@architect/functions)

> Runtime helper library for serverless apps built with [Architect][npm]

[![GitHub CI status](https://github.com/architect/functions/workflows/Node%20CI/badge.svg)](https://github.com/architect/functions/actions?query=workflow%3A%22Node+CI%22)
<!-- [![codecov](https://codecov.io/gh/architect/functions/branch/master/graph/badge.svg)](https://codecov.io/gh/architect/functions) -->

Check out the full docs: [arc.codes](https://arc.codes)


# API

Given:
```
let arc = require('@architect/functions')
```

The following APIs exists:


## `arc.events.subscribe(fn)`

Used to define a lambda function that will act as an [event][events] handler. Event handlers are defined in your application's Architect project manifest file under the [`@events`][events] pragma. The function code for the accompanying handler to each event should use `arc.events.subscribe` to wrap the handler. For example, given the following project manifest snippet:

```
@events
concerts
```

... the following file will be initialized representing the event handler for the `concerts` event, wherein you need to use `arc.events.subscribe`:

```
// file: src/events/concerts/index.js
let arc = require('@architect/functions')
module.exports = arc.events.subscribe(function(payload, callback) {
  console.log(payload)
  callback()
})
```


## `arc.events.publish(params, callback)`

Publishes `params.payload` to the SNS Topic (event) with name `params.name`. The `params.name` parameter should match the event defined under `@events`. Building on the example we described above, to trigger the `concerts` event handler, we would set `params.name` to be `concerts`.

This allows you to publish events from any function within your application (`@app` `.arc` file namespace) to be handled by the event handler.

When running in local/testing mode, will publish the event to the [sandbox][sandbox].


## `arc.queues.subscribe(params, callback)`

Used to define a lambda function that will act as a [queue][queues] handler. Queue handlers are defined in your application's `.arc` file under the [`@queues`][queues] pragma. The function code for the accompanying handler to each queued item should use `arc.queues.subscribe` to wrap the handler. For example, given the following `.arc` file snippet:

```
@queues
concert-tickets
```

... the following file will be initialized representing the event handler for the `concert-tickets` queue, wherein you need to use `arc.queues.subscribe`:

```
// file: src/queues/concert-tickets/index.js
let arc = require('@architect/functions')
module.exports = arc.queues.subscribe(function(payload, callback) {
  console.log(payload)
  callback()
})
```


## `arc.queues.publish(params, callback)`

Publishes `params.payload` to the SQS Queue (queue) with name `params.name`. The `params.name` parameter should match the queue defined under `@queues`. Building on the example we described above, to trigger the `concert-tickets` queue handler, we would set `params.name` to be `concert-tickets`.

This allows you to publish to queues from any function within your application (`@app` `.arc` file namespace) to be handled by the queue handler.

When running in local/testing mode, will publish the event to the [sandbox][sandbox].


## `arc.static(assetPath, options)`

Returns the fully-qualified URI of a static asset for the project-relative `assetPath` parameter. Takes into account:

- What environment (testing, staging, production) we are running in.
- Whether [fingerprinting][static] is enabled.
- Whether the override environment variable `ARC_STATIC_BUCKET` is present.

`options` is an object with the following currently-supported properties:

- `stagePath`: boolean, prepends `/staging` or `/production` to the asset path; useful if the current app is being run on an naked (non-domain-mapped) API Gateway

- `fullManifest`: boolean, returns an object with the full manifest of static assets instead of a single asset. `arc.static('',{fullManifest:true})` returns `{path:<path>,assets:[<asset1>,<asset2>]}`


## `arc.tables(callback)`

Returns an object that can be used to access data in database tables as defined under `@tables` in your `.arc` file. For example, given the following `.arc` file snippet:

```
@tables
accounts
  accountID *String

messages
  msgID *String
```

Running the following code:

```
let data = await arc.tables()
```

Would yield the following objects:

- `data.accounts`: reference to the `accounts` table
- `data.messages`: reference to the `messages` table

.. which contain the following methods:

- `delete(key, callback)`: deletes the record from the table with key `key` and invokes `callback` with the result
- `get(key, callback)`: retrieves the record from the table with key `key` and invokes `callback` when complete
- `put(item, callback)`: adds `item` to the table and invokes `callback` with the item when complete
- `query(params, callback)`: queries the table using `params` and invokes `callback` with the result
- `scan(params, callback)`: scans the table using `params` and invokes `callback` with the result
- `update(params, callback)`: updates an item in the table using `params` and invokes `callback` when complete


## `arc.ws.send({ id, payload })`

Sends the object present on `payload` to the connection ID on `id`. Payload is passed to `JSON.stringify()` on your behalf. Uses [`postToConnection`](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/ApiGatewayManagementApi.html#postToConnection-property) from the `ApiGatewayManagementApi`. Returns a promise with no data on success.

[npm]: https://www.npmjs.com/package/@architect/functions
[sandbox]: https://www.npmjs.com/package/@architect/sandbox
[events]: https://arc.codes/reference/events
[queues]: https://arc.codes/reference/queues
[static]: https://arc.codes/guides/static-assets


changelog.md
# Architect Functions changelog

---
Also see: [Architect changelog](https://github.com/architect/architect/blob/master/changelog.md)
---

## [3.15.0] 2021-07-02

### Added 
- New option for getting the full manifest of static assets from the arc.static() function. arc.static('',{fullManifest:true}) returns an object with the full list of files. This is useful for creating a cache manifest for a service worker etc.
---
## [3.14.1] 2021-05-25

### Fixed

- Fixed an issue in sandbox where table names containing other table names would
    get the table definitions confused; thanks @reconbot!

---

## [3.14.0] 2021-05-24

### Added

- New `arc.services()` API for retrieving the service map object for a given app; this object contains metadata associated with all infrastructure or services leveraged by the app

---

## [3.13.12] 2021-05-24

### Changed

- Removed unnecessary console logging for `@events` + `@queues` publishing

---

## [3.13.11] 2021-04-15

### Fixed

- Restored `req.path` convenience parameter missing from HTTP APIs; thanks @crtr0!

---

## [3.13.10] 2021-04-07

### Added

- Session cookie's SameSite value is configurable with ARC_SESSION_SAME_SITE environment variable; thanks @activescott!
- Fixed issue with `arc.tables()` not generating a client for tables with the string `production` in their names in a Sandbox context

---

## [3.13.9] 2021-01-13

### Added

- Added new alias for `arc.tables()._name` to `arc.tables().name`


### Fixed

- Fixed `arc.tables()._name` Sandbox mode returning an array instead of a string; thanks @exalted!

---

## [3.13.8] 2020-11-12

### Changed

- Internal change for JWE token issuance, thanks @lpetre!
- Ensure JWE session helper also uses the ordered last of multiple potential client sessions present in a cookie (if multiple are present)

---

## [3.13.7] 2020-11-05

### Changed

- Sessions helper now uses the ordered last of multiple potential client sessions present in a cookie (if multiple are present)


### Fixed

- Fixed subtle session bug where attempting to clear a session would retain old session data
- Fixed broken legacy `cookie` response param behavior, restoring behavior of being an alias for setting a `set-cookie` header, and not for writing a session

---

## [3.13.6] 2020-11-04

### Fixed

- Fixed rate-limit crashes in legacy `events.publish` call; fixes #300

---

## [3.13.5] 2020-11-03

### Changed

- Updated session test to no longer use legacy session table name
- Updated dependencies


### Fixed

- Handle body parsing `HTTP` APIs with v2.0 Lambda req payloads that include raw JSON; fixes @architect#997, thanks @mawdesley!

---

## [3.13.4] 2020-10-24

### Fixed

- Return errors on local `@events` + `@queues` publishing

---

## [3.13.3] 2020-09-23

### Fixed

- Fixes backward compat for legacy old school tidy Arc response params (e.g. `{ location: '/foo' }`) when used in `HTTP` APIs

---

## [3.13.2] 2020-09-21

### Fixed

- Fixes backward compat for legacy super old school content-type based responses (e.g. `{ html: 'foo' }`) when used in `HTTP` APIs

---

## [3.13.1] 2020-09-16

### Fixed

- Fixed bad call in `http.proxy` alias config

---

## [3.13.0] 2020-09-10

### Added

- Added full forwards compatibility for API Gateway HTTP APIs, including the new v2.0 payload
  - HTTP APIs in v2.0 have undocumented support for some funky response formats, so when in v2.0 mode, non-standard responses are passed through with minimal or no mutation
- Added support for passing along `multiValueHeaders` in responses

---

## [3.12.3] 2020-07-29

### Added

- Enable proxy to use HTTP APIs running in Lambda v1.0 payload format mode

---

## [3.12.2] 2020-07-20

### Added

- Adds support for loading compressed files out of proxy

---

## [3.12.1] 2020-06-16

### Fixed

- Fixed proxy lookup to custom 404 page working locally
- Fixed incorrect filename in proxy 404 error message

---

## [3.12.0] 2020-06-15

### Added

- Adds automatic fingerprint upgrading from non-fingerprinted requests; example:
  - If `@static fingerprint true` is enabled, `<img src=this-is-fine.gif>` will now automatically load `/_static/this-is-fine-abc123.gif`

---

## [3.11.0] 2020-06-07

### Added

- Adds fingerprinting support for pretty URLs and custom 404s in proxy
- Adds support for `ARC_STATIC_PREFIX` env var to provide symmetry with the new `@static prefix` setting
  - The `ARC_STATIC_FOLDER` env var is now considered deprecated, and will be removed in a future (breaking) release
- Added support for leading slashes in build-free templating (e.g. `${arc.static('/this-is-fine.gif')}`)


### Fixed

- Ensures the build-free templating engine always runs, not just when fingerprint is enabled

---

## [3.10.0 - 3.10.1] 2020-05-17

### Added

- Improved default caching behavior for static assets; fixes #273
  - Any assets that don't have `cache-control` explicitly set will now default to using `ETag` in order to improve the reliability of cache invalidation
  - HTML & JSON responses still default to anti-caching headers
- Added path peeking (aka pretty URLs); fixes #269
  - URLs delivered via `proxy` no longer require a trailing slash!
  - Example: to load `/foo/index.html`, you used to have to request `/foo/`
  - Now, if `/foo` is requested, `proxy` will automatically try peeking into `/foo` to see if it contains `/foo/index.html` – if not, it'll 404 as expected
- Added ETag support to Sandbox static asset serving
- Added support for arbitrary static asset config mapping
  - Pass proxy `config.assets` a `static.json`-like static asset object


### Changed

- Internal refactoring: cleaned up old `proxy` code paths


### Fixed

- We now ensure CDNs cannot cache `404` responses

---


## [3.9.1 - 3.9.3] 2020-05-06 - 2020-05-07

### Fixed

- `arc.http.express` POST request could fail if `req.body` was null

---

## [3.9.0] 2020-04-27

### Fixed

- `arc.http.express` POST request could fail if `req.body` was null


### Added


test/unit/src/static/index-test.js
let test = require('tape')
let proxyquire = require('proxyquire')
let env = process.env.NODE_ENV

let manifestExists = true
let fs = {
  readFileSync: () => (JSON.stringify({
    'foo.png': 'foo-1a2b3d.png'
  })),
  existsSync: () => manifestExists
}
let arcStatic = proxyquire('../../../../src/static', { fs })

function reset () {
  delete process.env.NODE_ENV
  if (process.env.NODE_ENV) throw ReferenceError('NODE_ENV not unset')
}

test('Set up env', t => {
  t.plan(1)
  t.ok(arcStatic, 'Static helper found')
})

test('Local env returns non-fingerprinted path', t => {
  t.plan(3)
  reset()
  manifestExists = true
  process.env.NODE_ENV = 'testing'
  let asset = arcStatic('foo.png')
  t.equal(asset, '/_static/foo.png', 'Returned non-fingerprinted path')
  asset = arcStatic('/foo.png')
  t.equal(asset, '/_static/foo.png', 'Returned non-fingerprinted path, stripping leading root slash')
  asset = arcStatic('foo.png', { stagePath: true })
  t.equal(asset, '/_static/foo.png', 'Returned non-fingerprinted path with stagePath option present')
})

test('Staging env returns _static path if root is requested', t => {
  t.plan(1)
  reset()
  manifestExists = false
  process.env.NODE_ENV = 'staging'
  let asset = arcStatic('/')
  t.equal(asset, '/_static/', 'Returned _static path')
})

test('Staging env returns non-fingerprinted path if static manifest is not present', t => {
  t.plan(1)
  reset()
  manifestExists = false
  process.env.NODE_ENV = 'staging'
  let asset = arcStatic('foo.png')
  t.equal(asset, '/_static/foo.png', 'Returned non-fingerprinted path')
})

test('Staging env returns fingerprinted path if static manifest is present', t => {
  t.plan(1)
  reset()
  manifestExists = true
  process.env.NODE_ENV = 'staging'
  let asset = arcStatic('foo.png')
  t.equal(asset, '/_static/foo-1a2b3d.png', 'Returned fingerprinted path')
})

test('Staging env returns non-fingerprinted path if file is not present in static manifest mapping', t => {
  t.plan(1)
  reset()
  manifestExists = true
  process.env.NODE_ENV = 'staging'
  let asset = arcStatic('bar.png')
  t.equal(asset, '/_static/bar.png', 'Returned non-fingerprinted path')
})

test('Passing stagePath option adds API Gateway /staging or /production to path', t => {
  t.plan(1)
  reset()
  manifestExists = true
  process.env.NODE_ENV = 'staging'
  let asset = arcStatic('foo.png', { stagePath: true })
  t.equal(asset, '/staging/_static/foo-1a2b3d.png', 'Returned fingerprinted path with API Gateway stage')
})

test('Reset', t => {
  reset()
  process.env.NODE_ENV = env
  t.end()
})


test/unit/src/http/proxy/format/templatize-test.js
let test = require('tape')
let { join } = require('path')

let filePath = join(process.cwd(), 'src', 'http', 'proxy', 'format', 'templatize')
let sut = require(filePath)

let buf = b => Buffer.from(b)

test('Module is present', t => {
  t.plan(1)
  t.ok(sut, 'Templatizer module is present')
})

test('Templatizer ignores binary responses', t => {
  t.plan(1)
  let content = 'here is an asset: ${STATIC(\'this-is-fine.gif\')}'
  let response = { body: buf(content) }
  let result = sut({
    isBinary: true,
    response
  })
  t.equal(content, result.body.toString(), 'Templatizer exited early')
})

test('Templatizer passes through non-fingerprinted assets', t => {
  t.plan(6)
  let response = { body: buf('here is an asset: ${STATIC(\'this-is-fine.gif\')}') }
  let result = sut({ response }).body.toString()
  t.notOk(result.includes('${STATIC(\'this-is-fine.gif\')}'), 'Templatizer stripped out STATIC')
  t.ok(result.includes('this-is-fine.gif'), 'Templatizer left in asset reference')

  response = { body: buf('here is an asset: ${arc.static(\'this-is-fine.gif\')}') }
  result = sut({ response }).body.toString()
  t.notOk(result.includes('${arc.static(\'this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes('this-is-fine.gif'), 'Templatizer left in asset reference')

  response = { body: buf('here is an asset: ${arc.static(\'/this-is-fine.gif\')}') }
  result = sut({ response }).body.toString()
  t.notOk(result.includes('${arc.static(\'/this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes('/this-is-fine.gif'), 'Templatizer left in asset reference (including leading slash)')
})

test('Templatizer replaces fingerprinted assets', t => {
  t.plan(6)
  let fingerprinted = 'this-is-fine-abc123.gif'
  let assets = {
    'this-is-fine.gif': fingerprinted
  }
  let response = { body: buf('here is an asset: ${STATIC(\'this-is-fine.gif\')}') }
  let result = sut({ response, assets }).body.toString()
  t.notOk(result.includes('${STATIC(\'this-is-fine.gif\')}'), 'Templatizer stripped out STATIC')
  t.ok(result.includes(fingerprinted), 'Templatizer replaced asset reference with fingerprint')

  response = { body: buf('here is an asset: ${arc.static(\'this-is-fine.gif\')}') }
  result = sut({ response, assets }).body.toString()
  t.notOk(result.includes('${arc.static(\'this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes(fingerprinted), 'Templatizer replaced asset reference with fingerprint')

  // Leading slash
  fingerprinted = '/this-is-fine-abc123.gif'
  response = { body: buf('here is an asset: ${arc.static(\'/this-is-fine.gif\')}') }
  result = sut({ response, assets }).body.toString()
  t.notOk(result.includes('${arc.static(\'/this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes(fingerprinted), 'Templatizer replaced asset reference with fingerprint (including leading slash)')
})

test('Templatizer does not replace fingerprinted assets locally', t => {
  t.plan(6)
  let assets = {
    'this-is-fine.gif': 'this-is-fine-abc123.gif'
  }
  let isLocal = true
  let response = { body: buf('here is an asset: ${STATIC(\'this-is-fine.gif\')}') }
  let result = sut({ response, assets, isLocal }).body.toString()
  t.notOk(result.includes('${STATIC(\'this-is-fine.gif\')}'), 'Templatizer stripped out STATIC')
  t.ok(result.includes('this-is-fine.gif'), 'Templatizer replaced asset reference with fingerprint')

  response = { body: buf('here is an asset: ${arc.static(\'this-is-fine.gif\')}') }
  result = sut({ response, assets, isLocal }).body.toString()
  t.notOk(result.includes('${arc.static(\'this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes('this-is-fine.gif'), 'Templatizer replaced asset reference with fingerprint')

  response = { body: buf('here is an asset: ${arc.static(\'/this-is-fine.gif\')}') }
  result = sut({ response, assets, isLocal }).body.toString()
  t.notOk(result.includes('${arc.static(\'/this-is-fine.gif\')}'), 'Templatizer stripped out arc.static')
  t.ok(result.includes('/this-is-fine.gif'), 'Templatizer replaced asset reference with fingerprint (including leading slash)')
})

test('can interpret js template strings as js', t=> {
  t.plan(1)
  let str = "<div>${ 'cat'.toUpperCase() + (1 + 1) }</div>"
  let response = { body: buf(str) }
  let result = sut({ response }).body.toString()
  t.ok(result === '<div>CAT2</div>', 'haha')
})


src/static/index.js
const { readFileSync, existsSync } = require('fs')
const { join } = require('path')

/**
 * Architect static asset helper
 * - Returns the live asset path and filename
 *
 * In order to keep this method sync, it does not use reflection to get fingerprint status
 * - Not checking @static fingerprint true (which we used to read from the .arc file) is possibly dangerous, so ensure asset path is valid
 * - ? TODO: add fingerprint state to env vars in Arc 6 to restore config safety?
 * @param {string} path - the path to the asset (eg. /index.js)
 * @returns {string} path - the resolved asset path (eg. /_static/index-xxx.js)
 */
module.exports = function _static (asset, options = {}) {
  let key = asset[0] === '/' ? asset.substring(1) : asset
  let isIndex = asset === '/'
  let manifest = join(process.cwd(), 'node_modules', '@architect', 'shared', 'static.json')
  let exists = existsSync(manifest)
  let local = process.env.NODE_ENV === 'testing' || process.env.ARC_LOCAL
  let stagePath = options.stagePath && !local ? '/' + process.env.NODE_ENV : ''
  let path = `${stagePath}/_static`
  if (!local && exists && !isIndex) {
    let read = p => readFileSync(p).toString()
    let pkg = JSON.parse(read(manifest))
    let asset = pkg[key]
    if (!asset)
      return `${path}/${key}`
    return `${path}/${asset}`
  }
  return `${path}/${isIndex ? '' : key}`
}


src/http/proxy/read/_local.js
let { existsSync, readFileSync } = require('fs')
let { extname, join, sep } = require('path')
let mime = require('mime-types')
let crypto = require('crypto')

let binaryTypes = require('../../helpers/binary-types')
let { httpError } = require('../../errors')
let transform = require('../format/transform') // Soon to be deprecated
let templatizeResponse = require('../format/templatize')
let normalizeResponse = require('../format/response')
let pretty = require('./_pretty')

/**
 * arc.http.proxy.read
 *
 * Reads a file from the local filesystem, resolving an HTTP Lambda friendly payload
 *
 * @param {Object} params
 * @param {String} params.Key
 * @param {String} params.IfNoneMatch
 * @param {String} params.isFolder
 * @param {String} params.isProxy
 * @param {Object} params.config
 * @returns {Object} {statusCode, headers, body}
 */
module.exports = async function readLocal (params) {

  let { ARC_SANDBOX_PATH_TO_STATIC, ARC_STATIC_PREFIX, ARC_STATIC_FOLDER } = process.env
  let { Key, IfNoneMatch, isFolder, isProxy, config, request } = params
  let headers = {}
  let response = {}

  // Unlike S3, handle basePath and assets inside the function as Sandbox is long-lived
  let staticAssets
  // After 6.x we can rely on this env var in sandbox
  let basePath = ARC_SANDBOX_PATH_TO_STATIC || join(process.cwd(), '..', '..', '..', 'public')
  let staticManifest = join(basePath, 'static.json')
  if (existsSync(staticManifest)) {
    staticAssets = JSON.parse(readFileSync(staticManifest))
  }
  let assets = config.assets || staticAssets

  // Look up the blob
  // Assume we're running from a lambda in src/**/* OR from vendored node_modules/@architect/sandbox
  let filePath = join(basePath, Key)
  // Denormalize static folder for local paths (not something we'd do in S3)
  let staticPrefix = ARC_STATIC_PREFIX || ARC_STATIC_FOLDER
  if (filePath.includes(staticPrefix)) {
    filePath = filePath.replace(`${staticPrefix}${sep}`, '')
  }

  try {
    // If client sends If-None-Match, use it in S3 getObject params
    let matchedETag = false

    // If the static asset manifest has the key, use that, otherwise fall back to the original Key
    let contentType = mime.contentType(extname(Key))

    if (!existsSync(filePath)) {
      let err = ReferenceError(`NoSuchKey: ${filePath} not found`)
      err.name = 'NoSuchKey'
      throw err
    }

    let body = readFileSync(filePath)
    let ETag = crypto.createHash('sha256').update(body).digest('hex')
    let result = {
      ContentType: contentType,
      ETag
    }
    if (IfNoneMatch === ETag) {
      matchedETag = true
      headers.ETag = IfNoneMatch
      response = {
        statusCode: 304,
        headers
      }
    }

    // No ETag found, return the blob
    if (!matchedETag) {
      let isBinary = binaryTypes.some(type => result.ContentType.includes(type) || contentType.includes(type))

      // Transform first to allow for any proxy plugin mutations
      response = transform({
        Key,
        config,
        isBinary,
        defaults: {
          headers,
          body
        }
      })

      // Handle templating
      response = templatizeResponse({
        isBinary,
        assets,
        request,
        response,
        isLocal: true
      })

      // Normalize response
      response = normalizeResponse({
        response,
        result,
        Key,
        isProxy,
        config
      })

      // Add ETag
      response.headers.ETag = result.ETag
    }

    if (!response.statusCode) {
      response.statusCode = 200
    }

    return response
  }
  catch (err) {
    let notFound = err.name === 'NoSuchKey'
    if (notFound) {
      return await pretty({ Key: filePath, config, isFolder })
    }
    else {
      let title = err.name
      let message = `
        ${err.message}<br>
        <pre>${err.stack}</pre>
      `
      return httpError({ statusCode: 500, title, message })
    }
  }
}


src/http/proxy/format/templatize.js
let tmpl = require('./lodash.template-4.5.0.js')

module.exports = function templatizeResponse (params) {

  let { isBinary, assets, request, response, isLocal=false } = params

  if (isBinary)
    return response

  // server scoped helper functions...
  let arc = {
    static(path) {
      let startsWithSlash = path.startsWith('/')
      let lookup = startsWithSlash ? path.substr(1) : path
      if (assets && assets[lookup] && !isLocal) {
        path = assets[lookup]
        path = startsWithSlash ? `/${path}` : path
      }
      return path
    },
    request
  }

  let body = response.body instanceof Buffer ? Buffer.from(response.body).toString() : response.body
  let replacer = tmpl(body) // TODO cache compiled function in a ledger keyed on ...somethin
  response.body = replacer({ arc, STATIC: arc.static }) // TODO Deprecate STATIC
  response.body = Buffer.from(response.body)
  return response
}


src/http/proxy/read/_s3.js
let { existsSync, readFileSync } = require('fs')
let { extname, join } = require('path')
let mime = require('mime-types')
let aws = require('aws-sdk')

let binaryTypes = require('../../helpers/binary-types')
let { httpError } = require('../../errors')
let transform = require('../format/transform') // Soon to be deprecated
let templatizeResponse = require('../format/templatize')
let normalizeResponse = require('../format/response')
let pretty = require('./_pretty')

/**
 * arc.http.proxy.read
 *
 * Reads a file from S3, resolving an HTTP Lambda friendly payload
 *
 * @param {Object} params
 * @param {String} params.Key
 * @param {String} params.Bucket
 * @param {String} params.IfNoneMatch
 * @param {String} params.isFolder
 * @param {String} params.isProxy
 * @param {Object} params.config
 * @returns {Object} {statusCode, headers, body}
 */
module.exports = async function readS3 (params) {

  let { Bucket, Key, IfNoneMatch, isFolder, isProxy, config, request } = params
  let { ARC_STATIC_PREFIX, ARC_STATIC_FOLDER } = process.env
  let prefix = ARC_STATIC_PREFIX || ARC_STATIC_FOLDER || config.bucket && config.bucket.folder
  let assets = config.assets || staticAssets
  let headers = {}
  let response = {}

  try {
    // If client sends If-None-Match, use it in S3 getObject params
    let matchedETag = false
    let s3 = new aws.S3

    // Try to interpolate HTML/JSON requests to fingerprinted filenames
    let contentType = mime.contentType(extname(Key))
    let capture = [ 'text/html', 'application/json' ]
    let isCaptured = capture.some(type => contentType.includes(type))
    if (assets && assets[Key] && isCaptured) {
      // Not necessary to flag response formatter for anti-caching
      // Those headers are already set in S3 file metadata
      Key = assets[Key]
    }

    // Check for progressive bundle upgrade request
    if (assets && assets[Key] && isCaptured === false) {
      return {
        statusCode: 302,
        headers: {
          location: `/${ assets[Key] }` 
        }
      }
    }

    /**
     * Folder prefix
     *   Enables a bucket folder at root to be specified
     */
    if (prefix) {
      Key = `${prefix}/${Key}`
    }

    let options = { Bucket, Key }
    if (IfNoneMatch) {
      options.IfNoneMatch = IfNoneMatch
    }

    let result = await s3.getObject(options).promise().catch(err => {
      // ETag matches (getObject error code of NotModified), so don't transit the whole file
      if (err.code === 'NotModified') {
        matchedETag = true
        headers.ETag = IfNoneMatch
        response = {
          statusCode: 304,
          headers
        }
      }
      else {
        // Important: do not swallow this error otherwise!
        throw err
      }
    })

    // No ETag found, return the blob
    if (!matchedETag) {
      let contentEncoding = result.ContentEncoding
      let isBinary = binaryTypes.some(type => result.ContentType.includes(type) || contentType.includes(type))

      // Transform first to allow for any proxy plugin mutations
      response = transform({
        Key,
        config,
        isBinary,
        defaults: {
          headers,
          body: result.Body
        }
      })

      // Handle templating
      response = templatizeResponse({
        isBinary,
        assets,
        request,
        response
      })

      // Normalize response
      response = normalizeResponse({
        response,
        result,
        Key,
        isProxy,
        config
      })

      // Add ETag
      response.headers.ETag = result.ETag
      // If encoded, add that too
      if (contentEncoding) response.headers['Content-Encoding'] = contentEncoding
    }

    if (!response.statusCode) {
      response.statusCode = 200
    }

    return response
  }
  catch (err) {
    let notFound = err.name === 'NoSuchKey'
    if (notFound) {
      return await pretty({ Bucket, Key, assets, headers, isFolder, prefix })
    }
    else {
      let title = err.name
      let message = `
        ${err.message}<br>
        <pre>${err.stack}</pre>
      `
      return httpError({ statusCode: 500, title, message })
    }
  }
}

/**
 * Fingerprinting manifest
 *   Load the manifest, try to hit the disk as infrequently as possible across invocations
 */
let staticAssets
let staticManifest = join(process.cwd(), 'node_modules', '@architect', 'shared', 'static.json')
if (staticAssets === false) {
  null /*noop*/
}
else if (existsSync(staticManifest) && !staticAssets) {
  staticAssets = JSON.parse(readFileSync(staticManifest))
}
else {
  staticAssets = false
}
